src/frontend/src/handler/alter_set_schema.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use pgwire::pg_response::StatementType;
+use risingwave_common::error::{ErrorCode, Result};
+use risingwave_pb::ddl_service::alter_set_schema_request::Object;
+use risingwave_sqlparser::ast::{ObjectName, OperateFunctionArg};
+
+use super::{HandlerArgs, RwPgResponse};
+use crate::catalog::root_catalog::SchemaPath;
+use crate::{bind_data_type, Binder};
+
+// Steps for validation:
+// 1. Check permission to alter original object.
+// 2. Check duplicate name in the new schema.
+// 3. Check permission to create in the new schema.
+
+/// Handle `ALTER [TABLE | [MATERIALIZED] VIEW | SOURCE | SINK | FUNCTION] <name> SET SCHEMA <schema_name>` statements.
+pub async fn handle_alter_set_schema(
+    handler_args: HandlerArgs,
+    obj_name: ObjectName,
+    new_schema_name: ObjectName,
+    stmt_type: StatementType,
+    func_args: Option<Vec<OperateFunctionArg>>,
+) -> Result<RwPgResponse> {
+    let session = handler_args.session;
+    let db_name = session.database();
+    let (schema_name, real_obj_name) =
+        Binder::resolve_schema_qualified_name(db_name, obj_name.clone())?;
+    let search_path = session.config().get_search_path();
+    let user_name = &session.auth_context().user_name;
+    let schema_path = SchemaPath::new(schema_name.as_deref(), &search_path, user_name);
+
+    let new_schema_name = Binder::resolve_schema_name(new_schema_name)?;
+    let object = {
+        let catalog_reader = session.env().catalog_reader().read_guard();
+
+        match stmt_type {
+            StatementType::ALTER_TABLE | StatementType::ALTER_MATERIALIZED_VIEW => {
+                let (table, old_schema_name) =
+                    catalog_reader.get_table_by_name(db_name, schema_path, &real_obj_name)?;
+                session.check_privilege_for_drop_alter(old_schema_name, &**table)?;
+                catalog_reader.check_relation_name_duplicated(
```

Comment:
The check will fail if the target schema is unchanged.

Reply:
FYI: if we try to set schema to the one that the object was already in in PostgreSQL, it will success silently.